### DISCUSSION TITLE
Replacing TypeORM with Drizzle ORM in Vendure

### DISCUSSION THREAD

**Discussion by asonnleitner** [vendurehq/vendure#2873] [General]
### Discussion Topic: Replacing TypeORM with Drizzle ORM in Vendure

Hello Vendure Community,

I would like to open a discussion regarding the replacement of [TypeORM](https://github.com/typeorm/typeorm) with the more modern [Drizzle ORM](https://github.com/drizzle-team/drizzle-orm) for Vendure's database interactions.

**Current Situation:**

TypeORM has been a go-to ORM, but it has shown several drawbacks, especially for larger projects. It can be slow, has become cumbersome to manage, and is not as actively maintained. 

**Proposal:**

I would like to suggest considering migrating to Drizzle ORM. Here are the key benefits:

- **Actively Maintained:** Drizzle ORM is a newer project and receives regular updates.
- **Lightweight & Edge Ready:** It's designed for high performance and low overhead.
- **Zero Dependencies:** Simplifies project dependencies and reduces potential conflicts.
- **Backed by Managed DB Providers:** Supported by modern and progressive DB providers like Turso, Xata, and Neon.
- **Wide Database Support:** Supports PostgreSQL, MySQL (including MariaDB), and SQLite.

**Advantages of Drizzle ORM:**

1. **Performance:** Drizzle ORM is optimized for top-notch performance, making it ideal for larger projects where TypeORM tends to be a bottleneck.
2. **Modern Features:** Leveraging features from managed database providers ensures we can use the latest advancements in database technology.
3. **Community and Support:** With a growing community and active maintenance, we can expect better support and faster issue resolution.

**Concerns and Considerations:**

- **Migration Effort:** Transitioning from TypeORM to Drizzle ORM will require significant effort in terms of code refactoring and testing.
- **Learning Curve:** Team members will need to familiarize themselves with Drizzle ORM's API and best practices.
- **Compatibility:** We need to ensure that Drizzle ORM's feature set covers all our current and future needs.

**Call to Action:**

I would love to hear the community's thoughts on this proposal. Have you experienced similar issues with TypeORM? Have you used Drizzle ORM, and if so, what has your experience been like? Let's discuss the potential challenges and benefits of making this switch.

Looking forward to a productive discussion!

**Comment by michaelbromley** [comment]
Thanks for your thoughtful proposal. 

Of all the TypeORM alternatives out there, Drizzle looks the most interesting to me. I'd be especially interested to hear anybody's experience using Drizzle on a very complex, high-data-volume project.

If we would ever choose to move away from TypeORM, the effort involved would be truly huge. Not just for us, but for everyone who has a significant amount of plugin code which inevitably touches lots of TypeORM APIs.

In fact, the amount of work, plus the inevitable fracturing of the ecosystem is such a huge risk that I'd consider forking TypeORM before going that route.

**Reply by asonnleitner** [reply]
Following up on this discussion after more time in production. The performance challenges with TypeORM continue to be a concern at scale.

@michaelbromley I've been thinking about the unique position Vendure is in now as the maintainer of TypeORM. While this gives direct control over the codebase, it also means Vendure is now responsible for maintaining a complex ORM for its entire ecosystem - a significant undertaking.

This raises an interesting question: Is maintaining TypeORM more resource-intensive than migrating to a more modern, actively developed ORM? Taking on TypeORM's maintenance means committing to its architecture and design decisions indefinitely.

Perhaps we should consider:
- Whether the TypeORM maintenance burden is sustainable long-term
- If resources spent maintaining TypeORM could be better invested elsewhere
- How to balance TypeORM's broad compatibility requirements with performance improvements
- Whether a gradual migration strategy might actually be less work than indefinite maintenance

The irony is that by becoming TypeORM's maintainer to solve our problems, we may have taken on an even larger challenge. Modern alternatives like Drizzle are actively developed by dedicated teams focused solely on ORM performance and features.

I'm curious about the community's thoughts on this. Is maintaining TypeORM the best path forward, or should we explore alternatives while we still have the flexibility to do so?

**Comment by jbeckton** [comment]
As an avid user of Drizzle in my other projects, I don't see how it can be extended, for example adding custom fields to the schema without actually modifying the schema. The current ORM allows for this, it's simple to add custom fields without having to modify core code.
